### PR TITLE
chore(internal): Refactor notepropsfnamedict to primitive typing

### PR DESCRIPTION
**Pull request by tma66**
**Background**
According to their official docs, Redux highly discourages use of non-serializable types (this includes classes and maps/sets). To bring parity between the redux state and storev2 state, we should use the same primitive javascript object typing.

In addition, `getNotesByFnameFromEngine` currently returns a reference to notes in the engine state. This means that any changes to the notes will affect the engine state. To prevent this side effect, we will return a copy instead.

**Major Changes**
1. Change type of `storev2.noteFnames` from NoteFNamesDict class object to NotePropsByFnameDict Object
2. Create `NoteFullDictUtils` class with static methods to work with both `NotePropsByFnameDict` and `NotePropsDict`. This will be used to add/update/delete from both dictionaries at same time
3. `getNotesByFnameFromEngine` will return a DEEP COPY instead of reference of notes from the engine. 
4. Update `DStore` api to include `getNote` and `findNotes`. These will be called by the engine to access notes from the store

**Minor Changes**
1. Move usages of `getNoteOrThrow` to `getNoteByFnameFromEngine` + throw themselves. getNoteOrThrow takes linear time to search
2. When DoctorService regenerates id, old note should be deleted
3. Minor bug fixes in store where we didn't keep track of all the NoteChangedEntry
4. Remove hydration step from `updateNote`. Notes should be hydrated on the client/engine side. This can lead to bugs where it removes parents/children from newNote if oldNote doesn't have.

# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](https://wiki.dendron.so/notes/773e0b5a-510f-4c21-acf4-2d1ab3ed741e.html)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](https://wiki.dendron.so/notes/pMS27sHxbWeKMoPRrWEzs.html).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics

- [ ] if you are adding analytics related changes, make sure you [Document](https://wiki.dendron.so/notes/8ThPaB9iXXm2Szk3C9kFt.html) changes in airtable

### Extended

- [ ] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 

### Basics

- [x] [Write Tests](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] [Confirm existing tests pass](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)
- [x] [Confirm manual testing](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [ ] If your tests changes an existing snapshot, snapshots have been [updated](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)

### Extended

- [ ] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](https://wiki.dendron.so/notes/dtMsF12SF2SUhLN10sYe2.html)

- CSS
  - [ ] display is correct for following dimensions
    - [ ] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [ ] lg: screen ≥ 992px
    - [ ] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [ ] display is correct for following browsers (across the various dimensions)
    - [ ] safari
    - [ ] firefox
    - [ ] chrome

### Basics

- [ ] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 

### Basics

- [x] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](https://wiki.dendron.so/notes/3489b652-cd0e-4ac8-a734-08094dc043eb.html) or [Packages](https://wiki.dendron.so/notes/32cdd4aa-d9f6-4582-8d0c-07f64a00299b.html)

## 

### Extended

- [ ]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](https://wiki.dendron.so/notes/iZ8vpfY0n1E3Nq3IC1Y7X.html)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

File: packages/common-all/src/types/typesv2.ts [hunk 1]
```diff
@@ -169,10 +170,25 @@ export type DNodePropsDict = {
...
 };
 
+/**
+ * Map of noteFname -> list of noteIds. Since fname is not unique across vaults, there can be multiple ids with the same fname
+ */
+export type NotePropsByFnameDict = {
+  [key: string]: string[];
+};
+
+export type NotePropsFullDict = {
```
Comment (kevinslin) [inline packages/common-all/src/types/typesv2.ts:187]: name is a little obtuse, especially since this doesn't return `NoteProps` by a dictionary instead. rename to `NoteDicts` instead?

File: packages/common-all/src/notePropDict.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  NotePropsFullDict,
+} from "./types";
+import { cleanName, isNotUndefined } from "./utils";
+import { VaultUtils } from "./vault";
+
+/**
+ * Utilities for working with NotePropsFullDict. The reason NotePropsFullDict is not a class is due to needing
+ * to work with primitive objects with redux
+ */
+export class NoteFullDictUtils {
```
Comment (kevinslin) [inline packages/common-all/src/notePropDict.ts:16]: rename to NoteDictsUtils

File: packages/common-all/src/notePropDict.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+   * Return empty array if no match is found
+   *
+   * @param fname
+   * @param notesDict
+   * @param vault If provided, use to filter results
+   * @returns Copy of NoteProps array
+   */
+  static findNotesByFname(
+    fname: string,
+    notesDict: NotePropsFullDict,
```
[the remark this answers is not in the document]
Comment (kevinslin) [reply packages/common-all/src/notePropDict.ts:28]: also apply to subsequent methods

File: packages/common-all/src/notePropDict.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    return _.cloneDeep(notes);
+  }
+
+  /**
+   * Remove note from both notesById and notesByFname
+   *
+   * @param note note to delete
+   * @param notesDict
+   */
+  static deleteNote(note: NoteProps, notesDict: NotePropsFullDict) {
```
Comment (kevinslin) [inline packages/common-all/src/notePropDict.ts:50]: should we return an error if the key isn't present?

File: packages/common-all/src/notePropDict.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+ * Utilities for working with NotePropsByFnameDict.
+ */
+export class NoteFnameDictUtils {
+  /**
+   * Create a map of {key -> value} where key = noteFname and value is list of ids corresponding to that fname from NotePropsDict
+   *
+   * @param notesById NotePropsDict used to populate map
+   * @returns
+   */
+  static create(notesById: NotePropsDict): NotePropsByFnameDict {
```
Comment (kevinslin) [inline packages/common-all/src/notePropDict.ts:93]: this might be a big refactor but for consistency, can we rename this to `NotePropsByIdDict`?

File: packages/engine-test-utils/src/presets/engine-server/rename.ts [hunk 6]
```diff
@@ -1599,7 +1599,7 @@ const NOTES = {
           expected: "foo1",
         },
         {
-          actual: changedEntries && changedEntries.length === 5,
+          actual: changedEntries && changedEntries.length === 6,
```
Comment (kevinslin) [inline packages/engine-test-utils/src/presets/engine-server/rename.ts:1602]: curiosu - why did this change?

**Review by kevinslin — CHANGES_REQUESTED** [review]
nice job! hopefully this should drive down the corner cases and engine being out of sync issues. left comments, mostly around conventions and naming

File: packages/common-all/src/types/typesv2.ts (hunk 1 above)
Comment (SeriousBug) [reply packages/common-all/src/types/typesv2.ts:187]: I'd also recommend a docstring that links to the utils class to remind anyone using this that they should be accessing through that class.

File: packages/engine-server/src/drivers/file/noteParser.ts [hunk 7]
```diff
@@ -214,21 +218,22 @@ export class NoteParser extends ParserBase {
                 hash: resp.noteHash,
               });
             }
+
+            // Check for duplicate IDs when adding new note to the map
+            if (notesById[notes[0].id] !== undefined) {
```
Comment (SeriousBug) [inline packages/engine-server/src/drivers/file/noteParser.ts:223]: The duplicate note ID check is now only happening for the first note in `notes`, is there any reason for that?

File: packages/engine-server/src/doctor/service.ts [hunk 8]
```diff
@@ -380,7 +380,8 @@ export class DoctorService implements Disposable {
           note.id = genUUID();
           await engine.writeNote(note, {
             runHooks: false,
-            updateExisting: true,
+            // Old note needs to be removed
```
Comment (jonathanyeung) [inline packages/engine-server/src/doctor/service.ts:383]: Is this change related to this PR?

**Review by jonathanyeung — CHANGES_REQUESTED** [review]

**Comment by jonathanyeung** [comment]
Can you also go ahead and add an `@deprecated` tag to `NoteUtils.getNoteOrThrow()`?

File: packages/common-all/src/types/typesv2.ts (hunk 1 above)
Comment (tma66) [reply packages/common-all/src/types/typesv2.ts:187]: Fixed

File: packages/common-all/src/notePropDict.ts (hunk 4 above)
Comment (tma66) [reply packages/common-all/src/notePropDict.ts:50]: Looks like its fairly standard to return a boolean instead to denote whether it was deleted or it. I can return that instead. We can let the clients construct the error (eg 404) if needed

File: packages/common-all/src/notePropDict.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  /**
+   * Find notes by fname. If vault is provided, filter results so that only notes with matching vault is returned
+   * Return empty array if no match is found
+   *
+   * @param fname
+   * @param notesDict
+   * @param vault If provided, use to filter results
+   * @returns Copy of NoteProps array
+   */
+  static findNotesByFname(
```
[the remark this answers is not in the document]
Comment (tma66) [reply packages/common-all/src/notePropDict.ts:26]: sure

File: packages/engine-test-utils/src/presets/engine-server/rename.ts (hunk 6 above)
Comment (tma66) [reply packages/engine-test-utils/src/presets/engine-server/rename.ts:1602]: Previously we were missing the changeEntry where root.children got updated from "[foo]" to "[]" before foo got deleted.

In the diff, I added a couple places where we forgot to propagate all the changedEntries.

File: packages/engine-server/src/drivers/file/noteParser.ts (hunk 7 above)
Comment (tma66) [reply packages/engine-server/src/drivers/file/noteParser.ts:223]: `resp.propsList` now returns the note we are trying to create + all the parents that were updated. I wrote this with the idea that only the first note should matter for duplicate check but now that I think about it, its possible that stubs could also have duplicate ids. Let me rework this.

File: packages/engine-server/src/doctor/service.ts (hunk 8 above)
Comment (tma66) [reply packages/engine-server/src/doctor/service.ts:383]: Needed to make this change to make one of the tests to pass which expected note ids to change.
In addition, before this change, the old note would still exist in the engine.

**Review by SeriousBug — APPROVED** [review]

**Review by jonathanyeung — APPROVED** [review]

**Review by kevinslin — APPROVED** [review]
